sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNullAwareHashJoinExec.scala

```diff
@@ -0,0 +1,205 @@
+          s"BroadcastNullAwareHashJoinExec only supports (LeftAnti + BuildRight) for now.")
+    }
+
+    val numOutputRows = longMetric("numOutputRows")
+    resultRdd.mapPartitionsWithIndexInternal { (index, iter) =>
+      val resultProj = genResultProjection
+      resultProj.initialize(index)
+      iter.map { r =>
+        numOutputRows += 1
+        resultProj(r)
+      }
+    }
+  }
+
+  override def needCopyResult: Boolean =
+    streamed.asInstanceOf[CodegenSupport].needCopyResult
+
+  override def inputRDDs(): Seq[RDD[InternalRow]] = {
+    streamed.asInstanceOf[CodegenSupport].inputRDDs()
+  }
+
+  override def doProduce(ctx: CodegenContext): String = {
+    streamed.asInstanceOf[CodegenSupport].produce(ctx, this)
+  }
+
+  private def genStreamSideJoinKey(
+      ctx: CodegenContext,
+      input: Seq[ExprCode]): (ExprCode, String) = {
+    ctx.currentVars = input
+    if (leftKeys.length == 1 && leftKeys.head.dataType == LongType) {
+      // generate the join key as Long
+      val ev =
+        BindReferences.bindReference[Expression](leftKeys.head, left.output).genCode(ctx)
+      (ev, ev.isNull)
+    } else {
+      // generate the join key as UnsafeRow
+      val ev = GenerateUnsafeProjection.createCode(ctx,
+        BindReferences.bindReferences[Expression](leftKeys, left.output))
+      (ev, s"${ev.value}.anyNull()")
+    }
+  }
+
+  override def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String = {
+    val relation: HashedRelation = prepareBroadcastHashedRelation
+    val broadcastRef = ctx.addReferenceObj("broadcast", relation)
+    val clsName = relation.getClass.getName
+
+    // Inline mutable state since not many join operations in a task
+    val relationTerm = ctx.addMutableState(clsName, "relation",
+      v => s"""
+              | $v = (($clsName) $broadcastRef).asReadOnlyCopy();
+              | incPeakExecutionMemory($v.estimatedSize());
+            """.stripMargin, forceInline = true)
+
+    val (keyEv, anyNull) = genStreamSideJoinKey(ctx, input)
+    val matched = ctx.freshName("matched")
+    val numOutput = metricTerm(ctx, "numOutputRows")
+    val found = ctx.freshName("found")
+
+    if (relation.inputEmpty) {
+      s"""
+         |// singleColumn NAAJ inputEmpty(true) accept all
+         |$numOutput.add(1);
+         |${consume(ctx, input)}
+       """.stripMargin
+    } else if (relation.anyNullKeyExists) {
+      s"""
+         |// singleColumn NAAJ inputEmpty(false) anyNullKeyExists(true) reject all
+       """.stripMargin
+    } else {
+      s"""
+         |// singleColumn NAAJ inputEmpty(false) anyNullKeyExists(false)
+         |boolean $found = false;
+         |// generate join key for stream side
+         |${keyEv.code}
+         |// Check if the key has nulls.
+         |if (!($anyNull)) {
+         |  // Check if the HashedRelation exists.
+         |  UnsafeRow $matched = (UnsafeRow)$relationTerm.getValue(${keyEv.value});
```

Comment:
So the unique key code path optimization of BHJ isn't relevant here ?

Reply:
isUniqueKeyCodePath is not relevant in NAAJ. because in NAAJ there is no other condition to check as in BHJ. as follow:

no matter isUniqueKey or not, if hashedRelation.getValue return a not null match, we dropped the row, else we keep the row. 

```
s"""
         |boolean $found = false;
         |// generate join key for stream side
         |${keyEv.code}
         |// Check if the key has nulls.
         |if (!($anyNull)) {
         |  // Check if the HashedRelation exists.
         |  $iteratorCls $matches = ($iteratorCls)$relationTerm.get(${keyEv.value});
         |  if ($matches != null) {
         |    // Evaluate the condition.
         |    while (!$found && $matches.hasNext()) {
         |      UnsafeRow $matched = (UnsafeRow) $matches.next();
         |      $checkCondition {   // IN NAAJ, this will be emtpy string
         |        $found = true;
         |      }
         |    }
         |  }
         |}
         |// special case for NullAwareAntiJoin, if anyNull in streamedRow, row should be dropped.
         |${ if (isNullAwareAntiJoin) s"else { $found = true; }" else ""}
         |if (!$found) {
         |  $numOutput.add(1);
         |  ${consume(ctx, input)}
         |}
       """.stripMargin
```